File: src/dsl.js
'use strict';

// 布局 DSL 词法/语法分析，产出 AST（语义检查在 compiler.js）。
//
// 语法概览：
//   struct Packet be {
//     magic:  u16 = 0xABCD;
//     hdr:    u16 { version: 4, type: 4, flags: 8 };
//     count:  u8;
//     items:  Item[count];
//     payload: bytes[length - 2];
//     body:   choose (hdr.type) { 1: u16; 2: bytes[2]; _: u8; };
//     tail:   u16 align(4);
//     crc:    crc16;
//   }

const { parseExpr } = require('./expr');

class DslError extends Error {
  constructor(message, line) {
    super(line ? `第 ${line} 行: ${message}` : message);
    this.name = 'DslError';
    this.line = line || 0;
  }
}

const INT_TYPES = new Set(['u8', 'u16', 'u32', 'i8', 'i16', 'i32']);
const CHECKSUMS = new Set(['crc16', 'crc32', 'sum8', 'sum16', 'xor8']);

function tokenize(src) {
  const toks = [];
  let i = 0;
  let line = 1;
  while (i < src.length) {
    const c = src[i];
    if (c === '\n') { line++; i++; continue; }
    if (/\s/.test(c)) { i++; continue; }
    if (src.startsWith('//', i)) {
      while (i < src.length && src[i] !== '\n') i++;
      continue;
    }
    if (src.startsWith('/*', i)) {
      const end = src.indexOf('*/', i + 2);
      if (end === -1) throw new DslError('块注释未闭合', line);
      line += src.slice(i, end).split('\n').length - 1;
      i = end + 2;
      continue;
    }
    if (/[0-9]/.test(c)) {
      let j = i;
      if (src.startsWith('0x', i) || src.startsWith('0X', i)) {
        j = i + 2;
        while (j < src.length && /[0-9a-fA-F]/.test(src[j])) j++;
        if (j === i + 2) throw new DslError('非法十六进制字面量', line);
      } else {
        while (j < src.length && /[0-9]/.test(src[j])) j++;
      }
      toks.push({ t: 'num', v: Number(src.slice(i, j)), line });
      i = j;
      continue;
    }
    if (/[A-Za-z_]/.test(c)) {
      let j = i;
      while (j < src.length && /[A-Za-z0-9_]/.test(src[j])) j++;
      toks.push({ t: 'ident', v: src.slice(i, j), line });
      i = j;
      continue;
    }
    if (':;{}[](),=.*/%+-<>&|!~'.includes(c)) {
      toks.push({ t: 'punct', v: c, line });
      i++;
      continue;
    }
    throw new DslError(`非法字符 '${c}'`, line);
  }
  return toks;
}

// 读取平衡括号内的文本（用于表达式），start 指向开括号之后，返回 [文本, 消耗token数]
function readBalanced(toks, start, open, close) {
  let depth = 1;
  let j = start;
  for (; j < toks.length; j++) {
    const t = toks[j];
    if (t.t === 'punct' && t.v === open) depth++;
    else if (t.t === 'punct' && t.v === close) {
      depth--;
      if (depth === 0) break;
    }
  }
  if (j >= toks.length) throw new DslError(`'${open}' 未闭合`, toks[start] ? toks[start].line : 0);
  // 无空格拼接：多字符运算符（==、<< 等）在 DSL 分词时已被拆开，需原样拼回
  const text = toks.slice(start, j).map((t) => (t.t === 'num' ? String(t.v) : t.v)).join('');
  return [text, j - start + 1];
}

function parseSchema(src) {
  const toks = tokenize(src);
  let pos = 0;
  const peek = () => toks[pos];
  const line = () => (toks[pos] ? toks[pos].line : 0);
  const eatPunct = (p) => {
    const t = toks[pos];
    if (t && t.t === 'punct' && t.v === p) { pos++; return true; }
    return false;
  };
  const expectPunct = (p) => {
    if (!eatPunct(p)) throw new DslError(`应为 '${p}'`, line());
  };
  const expectIdent = (what) => {
    const t = toks[pos];
    if (!t || t.t !== 'ident') throw new DslError(`应为${what}`, line());
    pos++;
    return t.v;
  };
  const expectNum = () => {
    const t = toks[pos];
    if (!t || t.t !== 'num') throw new DslError('应为数字', line());
    pos++;
    return t.v;
  };
  // 读取 [ ... ] 或 ( ... ) 中的表达式
  const readExpr = (open, close) => {
    expectPunct(open);
    const [text, consumed] = readBalanced(toks, pos, open, close);
    pos += consumed;
    try {
      return parseExpr(text);
    } catch (e) {
      throw new DslError(`表达式错误: ${e.message}`, line());
    }
  };

  function parseBitFields() {
    // 消费 '{' name: bits (, name: bits)* ,? '}'
    expectPunct('{');
    const fields = [];
    for (;;) {
      if (eatPunct('}')) break;
      const name = expectIdent('位域名');
      expectPunct(':');
      const bits = expectNum();
      fields.push({ name, bits });
      if (eatPunct(',')) continue;
      expectPunct('}');
      break;
    }
    return fields;
  }

  function parseCaseType() {
    const t = peek();
    if (t && t.t === 'ident') {
      if (INT_TYPES.has(t.v)) {
        pos++;
        const bits = Number(t.v.slice(1));
        return { t: 'int', bits, signed: t.v[0] === 'i' };
      }
      if (t.v === 'bytes') {
        pos++;
        return { t: 'bytes', count: readExpr('[', ']') };
      }
      // 结构体引用
      pos++;
      return { t: 'struct', name: t.v };
    }
    throw new DslError('分支类型应为整数类型、bytes[N] 或结构体名', line());
  }

  function parsePrimary() {
    const t = peek();
    if (!t || t.t !== 'ident') throw new DslError('应为类型', line());
    const v = t.v;
    if (INT_TYPES.has(v)) {
      pos++;
      const bits = Number(v.slice(1));
      const signed = v[0] === 'i';
      if (peek() && peek().t === 'punct' && peek().v === '{') {
        if (signed) throw new DslError('位域容器必须是无符号类型', t.line);
        return { t: 'bits', bits, fields: parseBitFields() };
      }
      return { t: 'int', bits, signed };
    }
    if (v === 'bytes') {
      pos++;
      return { t: 'bytes', count: readExpr('[', ']') };
    }
    if (v === 'choose') {
      pos++;
      const expr = readExpr('(', ')');
      expectPunct('{');
      const cases = [];
      for (;;) {
        if (eatPunct('}')) break;
        let value = null; // null 表示默认分支 '_'
        const ct = peek();
        if (ct && ct.t === 'ident' && ct.v === '_') {
          pos++;
        } else if (ct && ct.t === 'num') {
          pos++;
          value = ct.v;
        } else {
          throw new DslError('分支标签应为整数或 _', line());
        }
        expectPunct(':');
        const type = parseCaseType();
        expectPunct(';');
        cases.push({ value, type });
      }
      return { t: 'choose', expr, cases };
    }
    if (CHECKSUMS.has(v)) {
      pos++;
      return { t: 'checksum', algo: v };
    }
    // 结构体引用
    pos++;
    return { t: 'struct', name: v };
  }

  function parseFieldType() {
    let type = parsePrimary();
    // 数组后缀：仅整数与结构体可构成数组
    if (peek() && peek().t === 'punct' && peek().v === '[') {
      if (type.t !== 'int' && type.t !== 'struct') {
        throw new DslError('只有整数或结构体可以构成数组', line());
      }
      type = { t: 'array', elem: type, count: readExpr('[', ']') };
    }
    // align(N)
    if (peek() && peek().t === 'ident' && peek().v === 'align') {
      pos++;
      expectPunct('(');
      const n = expectNum();
      expectPunct(')');
      type.align = n;
    }
    // = 常量（仅整数字段）
    if (peek() && peek().t === 'punct' && peek().v === '=') {
      pos++;
      const value = expectNum();
      if (type.t !== 'int') throw new DslError('只有整数字段可以声明常量', line());
      type.const = value;
    }
    return type;
  }

  const structs = [];
  while (pos < toks.length) {
    const t = peek();
    if (!t || t.t !== 'ident' || t.v !== 'struct') {
      throw new DslError("顶层只允许 'struct' 定义", line());
    }
    pos++;
    const name = expectIdent('结构体名');
    let endian = 'be';
    if (peek() && peek().t === 'ident' && (peek().v === 'be' || peek().v === 'le')) {
      endian = toks[pos].v;
      pos++;
    }
    expectPunct('{');
    const fields = [];
    for (;;) {
      if (eatPunct('}')) break;
      const fname = expectIdent('字段名');
      expectPunct(':');
      const type = parseFieldType();
      expectPunct(';');
      fields.push({ name: fname, type, line: t.line });
    }
    structs.push({ name, endian, fields });
  }
  if (structs.length === 0) throw new DslError('至少需要一个 struct 定义', 0);
  return { structs };
}

module.exports = { DslError, parseSchema, INT_TYPES, CHECKSUMS };

File: backend/app/dsl.py
"""布局 DSL：词法 + 递归下降语法分析，产出未校验的 AST。

语法概览：

    endian be;

    struct Header {
        magic: u16;                 // 默认端序
        ver:   u8;
        flags: bits be { mode: 2; enc: 1; rsv: 5; };
    }

    struct Packet {
        hdr: Header;                // 嵌套结构
        length: u8;
        kind: u8;
        payload: choice (kind) {    // 联合分支
            case 0: { text: bytes[length]; }
            case 1: { n: u8; samples: [u16le; n]; }
            default: { raw: bytes[2]; }
        };
        align(4);                   // 填充到 4 字节倍数
        skip(1);                    // 跳过 1 字节
        crc: checksum(u16, crc16, $start);
        note: u8 if length > 0;     // 条件字段
        let total = length + 1;     // 派生值（不占位）
    }
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any, Optional

from . import expr as E


class DSLSyntaxError(Exception):
    pass


# ---------------------------------------------------------------- AST

@dataclass
class IntSpecA:
    bits: int
    signed: bool
    endian: Optional[str]  # None = 继承布局默认


@dataclass
class BitItemA:
    name: str
    bits: int
    cond: Any = None


@dataclass
class FIntA:
    name: str
    spec: IntSpecA
    cond: Any = None


@dataclass
class FBitsA:
    name: str
    endian: Optional[str]
    items: list[BitItemA]
    cond: Any = None


@dataclass
class FArrayA:
    name: str
    elem: Any  # IntSpecA | str(结构名)
    count: Any
    cond: Any = None


@dataclass
class FBytesA:
    name: str
    count: Any
    cond: Any = None


@dataclass
class FStructA:
    name: str
    struct: str
    cond: Any = None


@dataclass
class FChoiceA:
    name: str
    tag: Any
    cases: list  # list[(int|None, list[字段])]，None 表示 default
    cond: Any = None


@dataclass
class FAlignA:
    multiple: Any


@dataclass
class FSkipA:
    count: Any


@dataclass
class FChecksumA:
    name: str
    spec: IntSpecA
    algo: str
    src: Optional[str]  # None 表示 $start
    cond: Any = None


@dataclass
class FLetA:
    name: str
    expr: Any


@dataclass
class StructDefA:
    name: str
    fields: list = field(default_factory=list)


@dataclass
class LayoutA:
    structs: list = field(default_factory=list)
    default_endian: str = "be"


# ---------------------------------------------------------------- 词法

_IDENT_RE = re.compile(r"[A-Za-z_$][A-Za-z0-9_$]*")
_NUM_RE = re.compile(r"0[xX][0-9a-fA-F]+|0[bB][01]+|[0-9]+")
_OPS = ("<<", ">>", "<=", ">=", "==", "!=", "&&", "||",
        "{", "}", "(", ")", "[", "]", ":", ";", ",", ".", "?",
        "+", "-", "*", "/", "%", "<", ">", "=", "!", "~", "&", "|", "^")


@dataclass
class Tok:
    kind: str  # 'IDENT' | 'NUM' | 'OP' | 'EOF'
    text: str
    line: int
    col: int


def tokenize(src: str) -> list[Tok]:
    toks: list[Tok] = []
    i, line, col = 0, 1, 1
    n = len(src)
    while i < n:
        ch = src[i]
        if ch in " \t\r":
            i += 1
            col += 1
            continue
        if ch == "\n":
            i += 1
            line += 1
            col = 1
            continue
        if ch == "#" or src.startswith("//", i):
            j = src.find("\n", i)
            if j == -1:
                break
            i = j
            continue
        m = _IDENT_RE.match(src, i)
        if m:
            toks.append(Tok("IDENT", m.group(0), line, col))
            col += m.end() - i
            i = m.end()
            continue
        m = _NUM_RE.match(src, i)
        if m:
            toks.append(Tok("NUM", m.group(0), line, col))
            col += m.end() - i
            i = m.end()
            continue
        for op in _OPS:
            if src.startswith(op, i):
                toks.append(Tok("OP", op, line, col))
                i += len(op)
                col += len(op)
                break
        else:
            raise DSLSyntaxError(f"{line}:{col}: 无法识别的字符 {ch!r}")
    toks.append(Tok("EOF", "", line, col))
    return toks


def _num_value(text: str) -> int:
    if text.lower().startswith("0x"):
        return int(text, 16)
    if text.lower().startswith("0b"):
        return int(text, 2)
    return int(text, 10)


# ---------------------------------------------------------------- 语法

_INTTYPE_RE = re.compile(r"^([ui])([0-9]+)(le|be)?$")


class Parser:
    def __init__(self, toks: list[Tok]):
        self.toks = toks
        self.i = 0

    # ---- 基础 ----
    def peek(self, ahead: int = 0) -> Tok:
        return self.toks[min(self.i + ahead, len(self.toks) - 1)]

    def next(self) -> Tok:
        t = self.toks[self.i]
        if t.kind != "EOF":
            self.i += 1
        return t

    def err(self, msg: str, tok: Optional[Tok] = None) -> DSLSyntaxError:
        t = tok or self.peek()
        return DSLSyntaxError(f"{t.line}:{t.col}: {msg}（临近 {t.text!r}）")

    def expect_op(self, op: str) -> Tok:
        t = self.peek()
        if t.kind == "OP" and t.text == op:
            return self.next()
        raise self.err(f"期望 '{op}'")

    def accept_op(self, op: str) -> bool:
        t = self.peek()
        if t.kind == "OP" and t.text == op:
            self.next()
            return True
        return False

    def expect_ident(self, what: str = "标识符") -> str:
        t = self.peek()
        if t.kind == "IDENT":
            self.next()
            return t.text
        raise self.err(f"期望{what}")

    def expect_num(self) -> int:
        t = self.peek()
        if t.kind == "NUM":
            self.next()
            return _num_value(t.text)
        raise self.err("期望整数")

    def accept_kw(self, kw: str) -> bool:
        t = self.peek()
        if t.kind == "IDENT" and t.text == kw:
            self.next()
            return True
        return False

    def expect_kw(self, kw: str) -> None:
        if not self.accept_kw(kw):
            raise self.err(f"期望关键字 '{kw}'")

    # ---- 顶层 ----
    def parse_layout(self) -> LayoutA:
        lay = LayoutA()
        while self.peek().kind != "EOF":
            if self.accept_kw("endian"):
                t = self.peek()
                if t.kind == "IDENT" and t.text in ("le", "be"):
                    lay.default_endian = self.next().text
                else:
                    raise self.err("endian 后应为 le 或 be")
                self.expect_op(";")
            elif self.accept_kw("struct"):
                lay.structs.append(self.parse_struct())
            else:
                raise self.err("期望 'struct' 或 'endian'")
        return lay

    def parse_struct(self) -> StructDefA:
        name = self.expect_ident("结构名")
        self.expect_op("{")
        sd = StructDefA(name=name)
        while not self.accept_op("}"):
            if self.peek().kind == "EOF":
                raise self.err("结构体未闭合")
            sd.fields.append(self.parse_member())
        return sd

    def parse_member(self):
        if self.accept_kw("let"):
            name = self.expect_ident("变量名")
            self.expect_op("=")
            e = self.parse_expr()
            self.expect_op(";")
            return FLetA(name=name, expr=e)
        if self.accept_kw("align"):
            self.expect_op("(")
            e = self.parse_expr()
            self.expect_op(")")
            self.expect_op(";")
            return FAlignA(multiple=e)
        if self.accept_kw("skip"):
            self.expect_op("(")
            e = self.parse_expr()
            self.expect_op(")")
            self.expect_op(";")
            return FSkipA(count=e)
        name = self.expect_ident("字段名")
        self.expect_op(":")
        spec = self.parse_spec()
        cond = None
        if self.accept_kw("if"):
            cond = self.parse_expr()
        self.expect_op(";")
        spec.name = name
        spec.cond = cond
        return spec

    def parse_spec(self):
        t = self.peek()
        if t.kind == "IDENT":
            if t.text == "bits":
                return self.parse_bits()
            if t.text == "choice":
                return self.parse_choice()
            if t.text == "bytes":
                self.next()
                self.expect_op("[")
                e = self.parse_expr()
                self.expect_op("]")
                return FBytesA(name="", count=e)
            if t.text == "checksum":
                return self.parse_checksum()
            m = _INTTYPE_RE.match(t.text)
            if m:
                self.next()
                spec = IntSpecA(bits=int(m.group(2)),
                                signed=m.group(1) == "i",
                                endian=m.group(3))
                return FIntA(name="", spec=spec)
            # 嵌套结构引用
            self.next()
            return FStructA(name="", struct=t.text)
        if t.kind == "OP" and t.text == "[":
            self.next()
            et = self.peek()
            if et.kind != "IDENT":
                raise self.err("数组元素应为整数类型或结构名")
            m = _INTTYPE_RE.match(et.text)
            if m:
                self.next()
                elem: Any = IntSpecA(bits=int(m.group(2)),
                                     signed=m.group(1) == "i",
                                     endian=m.group(3))
            else:
                self.next()
                elem = et.text
            self.expect_op(";")
            count = self.parse_expr()
            self.expect_op("]")
            return FArrayA(name="", elem=elem, count=count)
        raise self.err("期望类型（整数/结构/bits/choice/bytes/checksum/数组）")

    def parse_bits(self) -> FBitsA:
        self.expect_kw("bits")
        endian = None
        t = self.peek()
        if t.kind == "IDENT" and t.text in ("le", "be"):
            endian = self.next().text
        self.expect_op("{")
        items: list[BitItemA] = []
        while not self.accept_op("}"):
            if self.peek().kind == "EOF":
                raise self.err("bits 块未闭合")
            nm = self.expect_ident("位域名")
            self.expect_op(":")
            w = self.expect_num()
            cond = None
            if self.accept_kw("if"):
                cond = self.parse_expr()
            self.expect_op(";")
            items.append(BitItemA(name=nm, bits=w, cond=cond))
        return FBitsA(name="", endian=endian, items=items)

    def parse_choice(self) -> FChoiceA:
        self.expect_kw("choice")
        self.expect_op("(")
        tag = self.parse_expr()
        self.expect_op(")")
        self.expect_op("{")
        cases: list = []
        seen_default = False
        while not self.accept_op("}"):
            if self.peek().kind == "EOF":
                raise self.err("choice 块未闭合")
            if self.accept_kw("case"):
                v = self.expect_num()
                self.expect_op(":")
                body = self.parse_case_body()
                cases.append((v, body))
            elif self.accept_kw("default"):
                if seen_default:
                    raise self.err("重复的 default 分支")
                seen_default = True
                self.expect_op(":")
                cases.append((None, self.parse_case_body()))
            else:
                raise self.err("期望 'case' 或 'default'")
        return FChoiceA(name="", tag=tag, cases=cases)

    def parse_case_body(self) -> list:
        self.expect_op("{")
        body: list = []
        while not self.accept_op("}"):
            if self.peek().kind == "EOF":
                raise self.err("case 体未闭合")
            body.append(self.parse_member())
        return body

    def parse_checksum(self) -> FChecksumA:
        self.expect_kw("checksum")
        self.expect_op("(")
        t = self.peek()
        m = _INTTYPE_RE.match(t.text) if t.kind == "IDENT" else None
        if not m:
            raise self.err("checksum 第一个参数应为整数类型，如 u16")
        self.next()
        spec = IntSpecA(bits=int(m.group(2)), signed=m.group(1) == "i",
                        endian=m.group(3))
        self.expect_op(",")
        algo = self.expect_ident("校验算法名")
        src = None
        if self.accept_op(","):
            src = self.expect_ident("校验起点（字段名或 $start）")
            if src == "$start":
                src = None
        self.expect_op(")")
        return FChecksumA(name="", spec=spec, algo=algo, src=src)

    # ---- 表达式（优先级递增） ----
    _BIN_LEVELS = [
        ("||",),
        ("&&",),
        ("|",),
        ("^",),
        ("&",),
        ("==", "!="),
        ("<", "<=", ">", ">="),
        ("<<", ">>"),
        ("+", "-"),
        ("*", "/", "%"),
    ]

    def parse_expr(self, level: int = 0):
        if level >= len(self._BIN_LEVELS):
            return self.parse_unary()
        left = self.parse_expr(level + 1)
        while True:
            t = self.peek()
            if t.kind == "OP" and t.text in self._BIN_LEVELS[level]:
                self.next()
                right = self.parse_expr(level + 1)
                left = E.Bin(t.text, left, right)
            else:
                break
        if level == 0 and self.accept_op("?"):
            t = self.parse_expr()
            self.expect_op(":")
            f = self.parse_expr()
            left = E.Cond(left, t, f)
        return left

    def parse_unary(self):
        t = self.peek()
        if t.kind == "OP" and t.text in ("-", "!", "~"):
            self.next()
            return E.Un(t.text, self.parse_unary())
        return self.parse_primary()

    def parse_primary(self):
        t = self.peek()
        if t.kind == "NUM":
            self.next()
            return E.Lit(_num_value(t.text))
        if t.kind == "OP" and t.text == "(":
            self.next()
            e = self.parse_expr()
            self.expect_op(")")
            return e
        if t.kind == "IDENT":
            if t.text == "true":
                self.next()
                return E.Lit(1)
            if t.text == "false":
                self.next()
                return E.Lit(0)
            if t.text == "len":
                self.next()
                self.expect_op("(")
                nm = self.expect_ident("len 的参数")
                self.expect_op(")")
                return E.Len(nm)
            self.next()
            path = [t.text]
            while self.accept_op("."):
                path.append(self.expect_ident("成员名"))
            return E.Ref(tuple(path))
        raise self.err("期望表达式")


def parse_layout(src: str) -> LayoutA:
    return Parser(tokenize(src)).parse_layout()
